File: data_structures/Python/graph.py
#Create an undirected graph class
class Graph():

	def __init__(self):
		self._nodes = {}

	def createNode(self, value):
		node = {}
		node[val] = value
		node["edges"] = {}
		self._nodes[value] = node

	def addEdge(self, fromNode, toNode):
		self._nodes[fromNode]["edges"] = toNode
		self._nodes[toNode]["edges"] = fromNode

	def hasEdge(self, fromNode, toNode):
		edges = self._nodes[fromNode]["edges"]
		if toNode in edges:
			return true
		else:
			return false

	def contains(self, value):
		if value in self._nodes:
			return true
		else:
			return false

	def removeNode(self, value):
		del self._nodes[value]

	def removeEdge(self, fromNode, toNode):
		del self._nodes[fromNode]["edges"][toNode]
		del self._nodes[toNode]["edges"][fromNode]

	def forEach(self, cb):
		for node in self._nodes:
			cb(self._nodes[node]["val"])

File: data_structures/JavaScript/graph.js
//Create an undirected (edges go both ways) graph with pseudoclassical instantiation
var Graph = function() {
  this.nodes = {};
}

Graph.prototype.createNode = function(value) {
  var node = {};
  node.val = value;
  node.edges = {};
  this.nodes[value] = node;
}

//Add an edge from one node to another
Graph.prototype.addEdge = function(fromNode, toNode) {
  this.nodes[fromNode].edges[toNode] = toNode;
  this.nodes[toNode].edges[fromNode] = fromNode;
}

//Check if the graph has an edge between two specified nodes
//We use "hasOwnProperty" to only check for user defined nodes
Graph.prototype.hasEdge = function(fromNode, toNode) {
  if (this.nodes[fromNode].hasOwnProperty(toNode)) {
    return true;
  }
  return false
}

//Check if a particular node is contained in the graph
Graph.prototype.contains = function(value) {
  if (this.nodes.hasOwnProperty(value)) {
    return true;
  }
  return false;
}

//Remove a node from the graph
Graph.prototype.removeNode = function(value) {
  delete this.nodes[value];
}

Graph.prototype.removeEdge = function(fromNode, toNode) {
  delete this.nodes[fromNode].edges[toNode];
  delete this.nodes[toNode].edges[fromNode];
}

Graph.prototype.forEach = function(cb) {
  for (var key in this.nodes) {
    if (this.nodes.hasOwnProperty(key)) {
      cb(this.nodes[key].val);
    }
  }
}

